### ISSUE TITLE
Metadata file 'notes.sqlite' causes sync conflict with multiple clients

### ISSUE THREAD

**Issue by brainchild0** [pbek/QOwnNotes#1111]
#### Expected behaviour

Multiple instances of client connect to same remote endpoint without any synchronization conflicts, except when the same note has been modified on both sides.

#### Actual behaviour

File `notes.sqlite` is reported as synchronization conflict.

#### Steps to reproduce

Configure an instances of QOwnNotes on each of two systems, to connect to the same location in the same Nextcloud instance.  Configure Nextcloud clients on each system also to synchronize those locations.

#### Suggestion

Store tag, trash, and related metadata in a separate configuration directory.  Also possible to use a hidden file file, although Nextcloud clients can be configured to synchronize hidden files, thereby reintroducing the same problem.

#### Workaround 

Per site, clients can be configured to exclude files based on name patterns.

**Comment by pbek** [comment]
Is there something that can be done in QOwnNotes? Seems like either a Nextcloud behavior or a problem with or a "maximum open file handler" problem where the operating system doesn't report that the file `notes.sqlite` or its folder was modified. QOwnNotes would reload the note folder if it was modified.

**Comment by pbek** [comment]
Maybe it has something to do that the database file is opened by QOwnNotes (Qt Sqlite Engine). On Windows I have a settings where I attempt to close it after operations, but that was quite experimental. A lot of bad stuff can happen there...

https://github.com/pbek/QOwnNotes/blob/14948f105b26c50dd3f62b97edec47a016af3545/src/services/databaseservice.cpp#L323-L338

**Comment by pbek** [comment]
Also see https://github.com/pbek/QOwnNotes/issues/926

**Comment by brainchild0** [comment]
I think the central issue is that the file contains the metadata for all notes, not just one note.  So for note contents, a conflict occurs only if the same note is modified in multiple locations, whereas the metadata file can conflict if any metadata is modified.  Also, boundary cases can occur that users may not anticipate, creating a lower quality of experience.  For example suppose a user configures a new QOwnNotes client and points it to an empty notes directory, 'then' configures the files synchronization cleint.  If a metadata file is already created in the otherwise empty location, then it is a conflict.  (I think that is what I observed, in my case.  If you think I am mistaken then I can try to reproduce.)

**Comment by pbek** [comment]
Maybe you can start by posting the rest of the requirements of the issue template to see where you were testing. :wink:

**Comment by brainchild0** [comment]
I could try to reproduce with the log pane open, but the fact is I never witnessed QOwnNotes misbehave or even complain.  The only bad thing that happened was the Nextcloud client complaining of a conflict.

**Comment by pbek** [comment]
Still, seeing the content of your *Debug settings* would be nice to help you... :smile:

**Comment by brainchild0** [comment]
I would like to provide further needed information, but I am not sure what to do.  I consider the problem I describe is independent of whether QOwnNotes is even running.  I don't know how to get QOwnNotes to report any problems, or why I would try.

Perhaps I did not adequately explain the problem.

If I configure the Nextcloud Desktop Client to synchronize a folder, and if both the client and server have files of the same name and location, but different data, from before the synchronization was configured, then the client reports a conflict.  Such is the behavior of the client, and I do not consider this behavior incorrect, because it is the only way the client can avoid data loss in general.

Thus, if I have a server that is already synchronized against a client, and that client has created a database file, and then I install the application on a second client, and configure QOwnNotes before configuring the Nextcloud file synchronization, then QOwnNote will create a new database, and the server and the client will find conflicting files of the same name at the time of the first synchronization attempt.  The client has a new database, and the server has preexisting one from the other client.  As synchronization has never occurred before, Nextcloud client has no means to resolve which version is new and which out of date.

QOwnNotes may be smart enough to work with other copies of itself, but the user experience will be that Nextcloud reports an error.

If you feel this issue is not worth resolving, then I understand, but at least it is now reported.

**Comment by pbek** [comment]
> Still, seeing the content of your Debug settings would be nice to help you... smile

Open the settings, click on *Debug*

**Comment by brainchild0** [comment]
I reproduced, and created a log recording the events surrounding the point when the Nextcloud client reports the conflict.  I hope it is helpful to you.

[qon.log](https://github.com/pbek/QOwnNotes/files/2762263/qon.log)

**Comment by pbek** [comment]
Thank you, but nothing very unusual is happening there. And I still don't know anything about your operating system and configuration.

**Comment by brainchild0** [comment]
I agree, nothing unusual is happening.  My OS is Linux Mint.  I don't know what features of my configuration interest you, but your emphasizing those details suggests to me that we are not communicating effectively.

The effects I have described are design features of Nextcloud.  They are not OS-dependent.  It is clear that if the user creates a synchronization folder on the client, and the folders on the client and desktop at that time have different versions of the same file, then the client reports a conflict.  

This behavior is correct, but if conflicts with design of QOwnNotes.

**Comment by pbek** [comment]
I don't have this issue on KDE Neon or Manjaro with the Nextcloud sync client.

**Comment by brainchild0** [comment]
Once the synchronization is configured and active, then the issue would not occur.  It is a special case at the time of _configuring a new instance_, and even then only when QOwnNotes is configured _before_ the Nextcloud client is configured.  

Have you tried the steps in this sequence, then reviewed the Nextcloud client log?

**Comment by pbek** [comment]
There can only be tho things happening.

- a lock on the sqlite database file because Qt keeps it open (see #926)
- too many files open by inotify

Both are hard to get around.

**Comment by brainchild0** [comment]
I don't understand why you say there can only be two things happening.  How did you conclude that, especially when I explained exactly what is happening, and why it's happening, without invoking either of these cases?

Try an experiment.  Create a folder on your Nextcloud server, populate it with a file called test.txt, and write "Hello" in the body of the file.  Then create a folder on your desktop, add a file called test.txt (same name as on server), and write "Goodbye" in the body of the file.  Now configure your Nextcloud client to synchronize  the two folders.

Does the client report a conflict before finishing the synchronization?  It should.  Otherwise, which is correct?  "Hello"?  "Goodbye"?  The client can't know.  

I did this experiment earlier.  It took only a few minutes.

**Comment by pbek** [comment]
> I did this experiment earlier. It took only a few minutes.

I can, if you can tell me what it has to do with QOwnNotes. :laughing:

**Comment by pbek** [comment]
I'm sorry, I cannot change the design of how Qt works with sqlite databases and even bad things happened when I tried to close the database after queries (see #926).

**Comment by brainchild0** [comment]
The behavior occurs whether or not QOwnNotes is actively running.  It is not a question of whether a database connection or file handle is open.  It is a question of whether a version of the database is introduced into a shared folder that is not recognized by the client as a derivative of the one previously on the server.

Consider the following steps:
1. Create an empty directory called `<notesdir>`.
1. Create a clean installation of QOwnNotes.
2. Open the application and configure the notes location to `<notesdir>`.  (The application immediately creates `<notesdir>/notes.sqlite`.)
3. Close the application.
4. Configure a share with the Nextcloud client to synchronize `<notesdir>` with an existing notes repository on the server.
5. Synchronize the share.  __(Nextcloud client complains about the conflicting versions of `<notesdir>/notes.sqlite`).__

QOwnNotes is not running, but it has placed data on the file system in such a way that Nextcloud is unhappy.

**Comment by pbek** [comment]
Thank you for your description.

> Close the application.

are you sure it is closed (as in quit)?

> it has placed data on the file system in such a way that Nextcloud is unhappy

this is all done by Qt automatically and I have no idea why Nextcloud would complain about a file with no open file handler, I'll try to reproduce your issue if I have time at hand...

**Comment by brainchild0** [comment]
Two-way synchronization algorithms use various strategies, but they all take the following general form:

For each file that exists on both endpoints:
1. Since the last synchronization, has neither version, client or server, been modified?  If so, then do nothing.
1. Otherwise, has the server version been modified but not the client version?  If so, then overwrite the client version with the server version.
1. Otherwise, has the server version not been modified but the client version has been?  If so, then overwrite the server version with the client version.
1. Otherwise, have both the server and client versions been modified?  Is so, then a conflict is identified, because neither the server or client versions may be overwritten without losing changes.  (In some systems, no conflict is identified if the content of both files is identical.)

Algorithms use various strategies to detect whether a file has changed since the last synchronization, including timestamps and checksums,  They also record data about each file during a synchronization operation, for a reference of comparison during the next operation.

In the special case of the first synchronization of a paired folder, no data from a previous operation is available.  Therefore, any file that exists on both endpoints is considered to have changed, and to be in conflict (except if the files are identical, and the algorithm detects it).  

During the first synchronization, if the client has previously created a version of `notes.sqlite`, and a different version already exists  on the server (e.g. from other clients), then the system cannot resolve which version to keep on both sides and which to discard, so it reports a conflict.  

__This effect has no relation to whether an open file handle exists, or any other runtime state of the system, only how the files appear on the file system.__

You generally do not see this effect after the first synchronization completes successfully, because when a change occurs on one client, it is synchronized to the server, and then to other clients, whose versions have not changed since the last synchronization operation, and so may be overwritten without conflict.  Nextcloud desktop clients synchronize frequently, and your notes collection is likely modified only by you.  Meanwhile you can only work on one client at a time, and synchronization completes faster than you switch clients.

Also as stated previously, if you synchronize the folder pair on a client before configuring your application on that client, then your application finds the copy of `notes.sqlite` that already exists from the synchronization, and modifies that copy, rather than creating a new file.  Again, the client detects that the client version has changed, and that change is propagated to all endpoints faster than you can switch clients.

**Comment by gitowok** [comment]
Hi, I can confirm I'm also getting sync conflicts. I'm on various gentoo and ubuntu desktop pcs, running latest versions of QOwnNotes via NextCloud.   Is it safe to just set a crontab to rm notes.sqlite if i'm not interested in any features other than editing raw text files?  Is there any way to disable metadata and generation of this sqlite file?  Can we set the path of this file to be outside of the notes folder structure, for example ~/.qownnotes/notes.sqlite

**Comment by pbek** [comment]
The file holds information about tags (per note folder) and thus needs to be inside the note folder to be synced. If you delete it it will be generated again (which shouldn't be a problem)

**Comment by brainchild0** [comment]
If the file only holds tags and is not used by other applications, only QOwnNotes, then many users may not need it.  Perhaps a configuration option would be beneficial to disable creation of the file.  I did not see such an option in the preferences dialog box.

**Comment by pbek** [comment]
The trashed items are also listed in the database and the database is needed by QOwnNotes, so I cannot simple allow to "disable" it.

**Comment by brainchild0** [comment]
Maybe some users would want a toggle switch for  disabling synchronization of the file, either by making it hidden or placing it in an external directory.

**Comment by pbek** [comment]
> toggle switch for disabling synchronization of the file

they could do that in their sync client, like the Nextcloud sync client (QOwnNotes doesn't do any syncing)

but I would like it much better to find out what is causing the sync conflict... since the mechanic in QOwnNotes didn't change for a long time I only can think of some thing of a newer Qt version or some changes in the Nextcloud client.

**Comment by brainchild0** [comment]
Yes, a user could write a filter rule for the file name, but it would affect all shares.  It is cleaner if the application does not write the file to the notes directory.  Or if the file begins with `.`, then it is filtered automatically, on systems which consider such files to be hidden.

Reducing the incidence of conflicts could be achieved by synchronizing metadata for each note separately.  The current design makes this behavior difficult to implement.  However, the Nextcloud Notes app is under active development, and I understand that a RESTful API is planned.  I expect that tags and other fields  will be supported natively.  Until this interface is specified and implemented, major revisions to the design on the client side may not be a good use of resources.  Once the interface is available, I would assume it will be possible to synchronize metadata at a record or notes level of granularity  in the application logic.

If the problem in the original report appeared recently, it could  be because the desktop-client synchronization logic has matured.  The behavior I have described is typical of 2-way synchronization algorithms, from what I understand.

**Comment by pbek** [comment]
## 19.3.4
- made an attempt to remedy `notes.sqlite` sync issues with Nextcloud

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if anything changed with the syncing issue.

**Comment by brainchild0** [comment]
I can try.  What are you expecting the new release to do differently with respect to this issue?

**Comment by pbek** [comment]
see https://github.com/pbek/QOwnNotes/commit/c17235d7d5e9faf3987aaf134092ef0c3fb61b77, I tried to do some things (that shouldn't be necessary) to not get the file locked.

**Comment by brainchild0** [comment]
Yes, I am just not sure how this change would relate or could relate to any change in the behavior I described, which as I explained appears to be determined by the state of the files on the file system, not the lock status.  I have no indication that the NextCloud desktop client is aware of such locks, and as I also explained, the behavior reproduces even when NextCloud is not running.

**Comment by pbek** [comment]
Which states of the files on the file system did you experience?

**Comment by brainchild0** [comment]
Please review my earlier comments (e.g. 17 Jan), which explain synchronization algorithms.  File checksum and date are primarily the information used.

**Comment by pbek** [comment]
I'm sorry, I still didn't grasp most of the things you were talking about in your big post on 17 Jan :smile: 
But if it's about the way the sqlite file is created we can't do a lot of it since it's done by Qt directly.

**Comment by brainchild0** [comment]
Well, you can consider adding an option to place the file outside of the location where NextCloud operates (or to make the file hidden following platform-specific conventions).  This solution might help reduce problems for some users.

I'm not sure what to say about the earlier comments.  I tried to be as thorough as possible in that posting, as well as in others that clarified the distinction between file metadata and lock state.

**Comment by pbek** [comment]
> Well, you can consider adding an option to place the file outside of the location where NextCloud operates (or to make the file hidden following platform-specific conventions). This solution might help reduce problems for some users.

that wouldn't solve the problem, since the file is needed for QOwnNotes to operate and needs to be synced too for tagging and the trash to work. If anyone chooses not to sync it he/she can simply exclude it in the Nextcloud client directly.

**Comment by brainchild0** [comment]
Adding a filter rule to the NextCloud client may have unwanted side effects, because they apply globally.  The client does not allow exclusion of specific files in the client-side folder, only creation of global name-based filter rules.  Also this approach couples the user's Nextcloud configuration to the design choices of QOwnNotes, such as   the name of the database file.

**Comment by pbek** [comment]
> Adding a filter rule to the NextCloud client may have unwanted side effects, because they apply globally. The client does not allow exclusion of specific files in the client-side folder, only creation of global name-based filter rules. Also this approach couples the user's Nextcloud configuration to the design choices of QOwnNotes, such as the name of the database file.

that would be a feature request for the Nextcloud client

> There now is a new release, could you please test it and report if anything changed with the syncing issue.

between two Linux Clients where I previously had sync issues with the `notes.sqlite` I now don't have any with version 19.3.4, at least I can't reproduce any currently. the database gets synced even if both computers are running and QOwnNotes is active on both. it also get synced if only one was running and the other is woken up from sleep

**Comment by brainchild0** [comment]
> > Adding a filter rule to the NextCloud client may have unwanted side effects, because they apply globally. The client does not allow exclusion of specific files in the client-side folder, only creation of global name-based filter rules. Also this approach couples the user's Nextcloud configuration to the design choices of QOwnNotes, such as the name of the database file.
> 
> that would be a feature request for the Nextcloud client

Someone could request per-file exclusion in Nextcloud client, but it is unlikely to be adopted because it breaks Nextcloud design, which presents the server content as a master copy, of which each client may select a subset.  Thus, the server view may not be a subset of the client view.

However, that feature is not what I am suggesting currently,.  I am suggesting that QOwnNotes support an option to change the location or attributes of the database file such that it would not be subject to syncing in Nextcloud.  QOwnNotes is intended to sit above a Nextcloud client deployment, so features in QOwnNotes that align with NextCloud design would be a sensible approach.

**Comment by pbek** [comment]
I'm still not sure what features of QON are not in align with Nextcloud design. 

>  I am suggesting that QOwnNotes support an option to change the location or attributes of the database file such that it would not be subject to syncing in Nextcloud.

implementing that would be a lot of work and does not solve anything since the file needs to be synced

and what attributes are you talking of? xattr? there are none (just checked)

do you still get those sync problems with the current version?

**Comment by brainchild0** [comment]
> I'm still not sure what features of QON are not in align with Nextcloud design.

That the internal structure of QON-managed directories separates those files that the user wishes to synchronize from those he does not wish to synchronize, such that only files in the former category are in fact synchronized.  

> 
> > I am suggesting that QOwnNotes support an option to change the location or attributes of the database file such that it would not be subject to syncing in Nextcloud.
> 
> implementing that would be a lot of work and does not solve anything since the file needs to be synced
> 

I see.  The suggestion was meant to be simple.  The minimal approach would be a single checkbox that determines whether the file is called `notes.sqlite` versus `.notes.sqlite`.

> and what attributes are you talking of? xattr? there are none (just checked)
> 

"Attributes" was not meant so literally.  Primarily I meant file name, and/or hidden flag on platforms with support, such as Windows.

> do you still get those sync problems with the current version?

While in principle I could perform this test, I had already invested some effort in isolating the problem to specific issues, which I then explained in detail, and which do not relate to the changes applied in the new version.   While I am willing to help in ways that are likely to benefit the effort, I would prefer not to invest further effort in tests that cannot reasonably be expected to help.  

I previously verified that the problem occurs when QON is not running.   The hypothesis that the behavior would improve by merely changing the way QON uses locks while running is inconsistent with previous observations.  Testing this hypothesis does not seem to be an activity that is likely to help.

**Comment by pbek** [comment]
> isolating the problem to specific issues

the issue that the Nextcloud client seems to be unable to sync a file? I wonder why you still seek fault in QON (or trying to make me cripple it) if QON isn't even running in your test. If you can find out what is so special about the file on your machines then I (or Nextcloud) may be able to help you...

**Comment by brainchild0** [comment]
Nextcloud isn't failing to sync the file.  It is identifying a conflict because neither the server or client version of the file can be identified as the correct version that should overwrite the other.  In such cases, Nextcloud   writes the version on the server to a new file on the client, with a similar name, and reports the event in the log.  The user is expected to choose which version to use, or to create a new version by manually merging the content of the two.  This behavior is entirely within the design of Nextcloud.  It is an intrinsic limitation of merge algorithms that conflicts can occur.

Think of it like interacting with a remote Git repository.  If you attempt to push a branch that was updated on the remote side after your most recent fetch, then the merge will fail.  But it is not a case of Git behaving incorrectly, but rather of Git telling you that you are responsible for resolving the conflicts before the remote repository can be updated in a nondestructive way. 

The file in question is not "special".  The reason why the sync does not complete without a conflict is that in the particular case that I described, the file was modified in multiple places such that the conflict cannot be resolved automatically.

When a user has configured  Nextcloud to keep a directory in sync with a server, the user will normally avoid creating conflicts.  I suggest that QON similarly avoid creating conflicts with files it creates.  This suggestion is not crippling QOwnNotes, but rather operating in a way that is compatible with the correct, expected behavior of Nextcloud.

**Comment by pbek** [comment]
The only thing I can see so far is that I didn't get any sync conflict between any of my computers so far (3 Linux, 1 Mac, 1 Windows) since 19.3.4.

**Comment by brainchild0** [comment]
Interesting that you have such a result.

A few items to check:

1.  Originally, the local directory   contains no `notes.sqlite` file.
1.  The application then creates a `notes.sqlite` in that directory.
1.  The server already contains a `notes.sqlite` in some directory.
1.  Finally, after all of the above steps, the synchronization pair is  configured for the two directories, local and remote, each containing distinct `notes.sqlite`files, and the synchronization task is initialized.

If the Nextcloud client log   does not report a conflict, then are you able to determine whether either file, the one from the server or client, is overwriting the other?  Or do the files remain out of sync?

**Comment by pbek** [comment]
I don't understand steps 2 and 3. Did you use QON in a new note folder? Why has the server already a notes.sqlite?

**Comment by brainchild0** [comment]
> Did you use QON in a new note folder? 

Yes, this detail is on the list of steps to reproduce.

> Why has the server already a notes.sqlite?

Because the folder on the server was previously synchronized with a different desktop system that also runs QON.

**Comment by pbek** [comment]
> Because the folder on the server was previously synchronized with a different desktop system that also runs QON.

but why was it then locally empty for you? or did you only sync for the first time afterwards? if yes, what has this to do with QON?

**Comment by brainchild0** [comment]
Suppose you have a system called *Desktop 1* with successful synchronization with the server.

Now suppose you buy a new   system, called *Deskop 2*.  You install QON, set a notes folder, and add a new note `foo.md`.  QON creates a database file in the local folder, `notes.sqlite`.  The database file is different from the one of the same name shared by the server and *Desktoip 1*.

Now suppose you configure the notes folder in *Desktop 2* to sync against the server.  The new notes file is uploaded to the server without conflict, but the database file cannot be synced without a conflict because a file with the same name already exists on the server, and neither file is a newer version than the other.

**Comment by pbek** [comment]
That is why you're synchronizing your folders before you set up qownnotes.

brainchild0 <notifications@github.com> schrieb am Do., 28. März 2019, 17:36:

**Comment by brainchild0** [comment]
I don't understand.  

I stated the particular conditions under which the problem appears.   This is how bugs are reported.  If I had stated the conditions where no problem appears, then I would not successfully  be filing a  bug report.

**Comment by pbek** [comment]
As much as I enjoy writing with you, would you be willing to file a bug for the Nextcloud client if you experience problems with syncing files to Nextcloud? I don't think I'm able to help you with that as much as I like to.

**Comment by brainchild0** [comment]
There is no bug in Nextcloud that is causing this problem.

**Comment by chefschaffner** [comment]
Hi there, 
as I have the same problem, I try to clarify brainchild0's point.
I agree, it's not a problem of nextcloud. Furthermore I guess you will not be able to fix this problem for all cases. 
If you are syncing two devices and you change the same file at both devices before syncing, you will have the described problem. QOwnNotes cannot fix that, nor can Nextcloud.
Steps should be as this:
1. Sync the two devices
2. Change contents of an existing file "test.txt" at device one
3. Change contents of the same file "test.txt" at device two
4. Sync the two devices
5. described sync error will occur
As the sqlite-file from QON contains informations of all notes, this will happen if you change any note in the described way, because one file contains informations of all notes. 
If the sqlite file wouldn't exist, the conflict would only occur, if the **same** note was changed on both sides.  As the sqlite-file is changes every time you alter **any** note, the probability, that the sync-error occurs is much higher.
I could think of storing meta-informations in single files (one per note, e.g. Note is "test.txt", Metafile is "test.meta"), but I don't know, if that was possible for QON.
Hope I could help a little bit - I like QON very much (Using it on several clients, Windows, Linux & Android) and avoiding this sync-error would be a big enhancement in my eyes.

**Comment by pbek** [comment]
Thank you for your attempt to help.
`notes.sqlite` doesn't store any information but tags and tag-note-assignments. So the file will not be altered every time you edit a note, just if you add a tag to a note.

**Comment by brainchild0** [comment]
And of course it also appears in the case documented originally, when the first sync to Nextcloud occurs after the folder being designated by QON as the notes folder.

A new thought occurs to me, which might alleviate problems from this case.  Could the creation of the `notes.sql` file occur lazily, that is, only after the first tag is added?  Such a solution could have the ancillary benefit that a user not using tags need not carry the extra file in the notes folder.

**Comment by pbek** [comment]
The tag tree wouldn't work if there are no tag tables present. And there is other information stored too, like the trash...

**Comment by brainchild0** [comment]
> The tag tree wouldn't work if there are no tag tables present. 

I'm not sure what the "tag tree" is, or why it is needed before any notes are created.

> And there is other information stored too, like the trash...

Until the folder is populated with notes, there is no trash.  So this information can also be created lazily.  

Is there any information that cannot be deferred?  If a folder is vacant of notes, as every folder is until a note is first added, can it not quite suitably be be represented without the SQLite database?

**Comment by pbek** [comment]
> "tag tree"

The one in the user interface.

**Comment by pbek** [comment]
> can it not quite suitably be be represented without the SQLite database?

the user interface will not work without the database

**Comment by gitowok** [comment]
Heres my workaround:
rm notes.sqlite ; touch notes.sqlite ; sudo chown root:root notes.sqlite

This hard blocks QOwnNotes from accessing it. The user interface seems to work fine sofar. I dont use tag/trash functionality. I had to do this on all instances of QOwnNotes that try to access it, otherwise nextcloud would erase the file and recreate it with read/write permissions.

I wish there was a simple switch to turn off tags, trash and notes.sqlite entirely.

**Comment by pbek** [comment]
> I wish there was a simple switch to turn off tags, trash and notes.sqlite entirely.

too deeply ingrained to make that possible easily, I'd rather want the problem fixed

**Comment by pbek** [comment]
Meanwhile...

## 19.7.9
- every time the note folder is reloaded it will now be **checked** for
  **conflicting copies** of the **`notes.sqlite` database** that will block your
  Nextcloud/ownCloud sync process
    - you will be asked if you want to remove the conflicting copies

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if it works for you?

**Comment by pbek** [comment]
I will close this issue until there is more information.

**Comment by brainchild0** [comment]
Confirmed new behavior (tested commit 8a8ec98):
- Nextcloud client synchronization summary reports: `There are unresolved conflicts.`
- Nextcloud client log reports: 
   ~~~
   Conflict: Server version downloaded, local copy renamed and not uploaded.
   notes (conflicted copy YYYY-MM-DD 001111).sqlite
   ~~~
- QON application asks:
   ~~~
   Delete 1 conflicted database copies that may block your Nextcloud / ownCloud sync process?
   <notes folder>/notes (conflicted copy YYYY-MM-DD 001111).sqlite
   ~~~
- If above answered "Yes", then conflict resolves on next sync event.

**Comment by pbek** [comment]
Thank you for your review, @brainchild0. Is this solution satisfying enough for you?

**Comment by brainchild0** [comment]
I characterize this solution as satisfying.  

I think it is a good solution, considering the context and circumstances.  The quality of the user experience improves because the conflict can be promptly resolved automatically rather than needing to be resolved manually, or left unresolved.  As some early comments plainly imply, other overall application designs are conceivable that avoid the central problem and provide more dynamic and flexible behavior, but such is far beyond the scope of this issue.

One small note is that a user who is less knowledgeable might have difficulty deciding how to respond to the question.  The wording of the question suggests that an affirmative response is preferred, but it may be helpful to such a user if the phrasing more explicitly or forcefully indicates which answer is preferred in cases of doubt.  For this question, I suspect that it would be very rare that any user would ever answer in the negative.  The idea that an affirmative response is normal and a negative one exceptional can be reflected by wording such as, "Proceed with automatic deletion of conflict?".

**Comment by pbek** [comment]
> I think it is a good solution, considering the context and circumstances. The quality of the user experience improves because the conflict can be promptly resolved automatically rather than needing to be resolved manually, or left unresolved.

I also think it's the best solution, since I can't hope to find another way to get around the sync problem.

>  "Proceed with automatic deletion of conflict?".

Good idea, I'll do that.

**Comment by pbek** [comment]
## 19.8.2
- made confirmation message for conflicted database copies more affirmative
